esphome/components/json/json_util.cpp

```diff
@@ -23,21 +23,21 @@ std::string build_json(const json_build_t &f) {
 #ifdef USE_ESP8266
   const size_t free_heap = ESP.getMaxFreeBlockSize();  // NOLINT(readability-static-accessed-through-instance)
 #elif defined(USE_ESP32)
-  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_8BIT);
+  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_INTERNAL);
 #endif
 
-  const size_t request_size = std::min(free_heap, (size_t) 512);
+  const size_t request_size = std::min(free_heap - 2048, (size_t) 5120);
 
   DynamicJsonDocument json_document(request_size);
-  if (json_document.capacity() == 0) {
+  if (json_document.memoryPool().buffer() == nullptr) {
     ESP_LOGE(TAG, "Could not allocate memory for JSON document! Requested %u bytes, largest free heap block: %u bytes",
              request_size, free_heap);
     return "{}";
   }
   JsonObject root = json_document.to<JsonObject>();
   f(root);
   json_document.shrinkToFit();
-  ESP_LOGV(TAG, "Size after shrink %u bytes", json_document.capacity());
+
```

Comment:
```suggestion
ESP_LOGV(TAG, "Size after shrink %u bytes", json_document.capacity());
```
We don't need to revert this